src/sp800_108.rs

```diff
@@ -0,0 +1,396 @@
+// Copyright 2024 Simo Sorce
+// See LICENSE.txt file for terms
+
+fn key_to_segment_size(key: usize, segment: usize) -> usize {
+    ((key + segment - 1) / segment) * segment
+}
+
+const MAX8: usize = u8::MAX as usize;
+const MAX16: usize = u16::MAX as usize;
+const MAX24: usize = MAX8 * MAX16;
+const MAX32: usize = u32::MAX as usize;
+const MAX40: usize = MAX8 * MAX32;
+const MAX48: usize = MAX16 * MAX32;
+const MAX56: usize = MAX8 * MAX16 * MAX32;
+const MAX64: usize = u64::MAX as usize;
+
+fn ctr_update(
+    param: &Sp800CounterFormat,
+    ctr: usize,
+    op: &mut Box<dyn Sign>,
+) -> KResult<()> {
+    if !param.defined {
+        return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+    }
+    match param.bits {
+        8 => {
+            if ctr > MAX8 {
+                return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+            }
```

Comment:
See comment for DKLEN below... but for counters, its more likely that we don't want to overflow the counter because that could result in repeat blocks, so I wouldn't change the code here.

Reply:
I have a patch to change these errors into just truncating the value, will submit soon, thanks for spotting this.

Reply:
We can't overflow the counter per spec, the size of the counter is an explicit way to control how many iterations are allowed.